### PR TITLE
Fortran mangling and hiding blas headers. 

### PR DESCRIPTION
**Pull request by kyungjoo-kim**
This might impact the overall building procedure. KokkosKernels and Kokkos are independent from Trilinos and it also needs independent building systems. We support a few building system i.e., makefile, cmake and probably spack later. This PR is about fortran inter-operatibility in Makefile system. 

1. Problem

In KokkosKernels, we hardwire the fortran interface in tpl interface header. For instance, 
```
extern "C" void zdotc_(std::complex<double>*res, 
  int &N, 
  const std::complex<double>*a, int *inc_a, 
  const std::complex<double>*b, int *inc_b);

/// later we use this function insider tpl interface.
```
One problem is that fortran is tricky and sometimes it is callable by ``zdot``, ``zdot_``, ``zdot__`` and its upper cases. Also, some fortran interface requires a different interface e.g.,
```
extern "C" std::complex<double> zdotc_(int &N, 
  const std::complex<double>*a, int *inc_a, 
  const std::complex<double>*b, int *inc_b);
```
Let's say that this is the case ``blas-return-complex``.

We can do test these via some linking tests in cmake or other advanced build system. In my knowledge, I don't know how to test the fortran manling and link interface only using Makefile syntax.

2. Solution in Makefile.

For this moment, we just give manual overriding options to determine fortran mangling and complex interface. One can override fortran mangling and complex interface. An example is  
```
/// intel compilers, default options
../../master/scripts/generate_makefile.bash --with-options='eti-only,blas-mangle_'
/// this will link to void zdotc_(std::complex<double> *result, int &N ....) function

/// power8 gcc, 
../../master/scripts/generate_makefile.bash --with-options='eti-only,blas-mangle_,blas-return-complex'
/// this will link to "std::complex<double> zdotc_(int &N, ....)" function

/// power8 xl (I did not tested yet)
../../master/scripts/generate_makefile.bash --with-options='eti-only,blas-mangle,blas-return-complex'
/// this will link to "std::complex<double> zdotc(int &N, ....)" function
```

So, regarding the fortran mangling, we have three cases: ``blas-mangle,blas-mangle_,blas-mangle__``. Each option will add no underscore, single underscore and double underscores. 

``blas-return-complex`` will choose functions that can return complex values as a return type. 

3. Solution to exposed blas headers.

This has been @mhoemmen 's major complaint. Now it is hidden inside a KokkosBlas_Host_tpl.cpp file. 

@mhoemmen If I push this modification into Trilinos kokkoskernels, can you add your cmake modification of the linking test and squeeze the detected mangling format into KokkosKernels_config.h.in ?

### CODE REVIEW

**Comment by kyungjoo-kim** [comment]
referencing #369

**Comment by kyungjoo-kim** [comment]
@ndellingwood done. I only change blas-return-complex option in white test_all_sandia script as we use gcc compiled library for testing (it uses single underscore and it returns complex). I tested on white. It passes most of the tests and not sure if other tests can finish in time. Anyway, I think it is good to be merged. 
```
[kyukim @white22] spotcheck > . ../spotcheck.sh
Running on machine: white
netlib/3.8.0/gcc/7.2.0(5):ERROR:150: Module 'netlib/3.8.0/gcc/7.2.0' conflicts with the currently loaded module(s) 'netlib/3.8.0/ibm/xl/16.1.0'
netlib/3.8.0/gcc/7.2.0(5):ERROR:102: Tcl command execution failed: conflict       	netlib blas lapack openblas
Going to test compilers:  gcc/6.4.0 gcc/7.2.0 ibm/16.1.0 cuda/9.2.88
Testing compiler gcc/6.4.0
Testing compiler gcc/7.2.0
  Starting job gcc-6.4.0-OpenMP_Serial-release
  Starting job gcc-7.2.0-OpenMP-release
  PASSED gcc-7.2.0-OpenMP-release
  Starting job gcc-7.2.0-Serial-release
  PASSED gcc-6.4.0-OpenMP_Serial-release
Testing compiler ibm/16.1.0
  Starting job gcc-7.2.0-OpenMP_Serial-release
  PASSED gcc-7.2.0-Serial-release
Testing compiler cuda/9.2.88
  Starting job ibm-16.1.0-Serial-release
  PASSED gcc-7.2.0-OpenMP_Serial-release
```

**Comment by ndellingwood** [comment]
@kyungjoo-kim thanks! Based on the error message in your spotcheck output, it looks like a `module purge` should be called before running the test_all_sandia script, previously or auto- loaded modules likely caused the module conflict.

**Comment by ndellingwood** [comment]
@kyungjoo-kim also, are we sure the BLAS tpl code path is getting hit in the tests?

File: CMakeLists.txt [hunk 1]
```diff
@@ -323,6 +323,15 @@ IF (KOKKOSKERNELS_ENABLE_TPL_CUBLAS)
   LIST(APPEND TPL_LIST "CUBLAS")
 ENDIF()
 
+# ==================================================================
+# Fortran Complex BLAS
+# ==================================================================
+
+IF (KOKKOSKERNELS_INTERNAL_ENABLE_HOST_BLAS)
+  INCLUDE(CheckHostBlasReturnComplex.cmake)
```
Comment (ndellingwood) [inline CMakeLists.txt]: @jjwilke adding you here to keep your work on #120 in sync with a couple updates that were made for CMake builds (and Makefile builds) to support testing and setting proper fortran mangling and APIs for proper BLAS TPL support.

File: cmake/KokkosKernels_config.h.in [hunk 2]
```diff
@@ -1,6 +1,15 @@
...
 #define KOKKOSKERNELS_CONFIG_H
 
+
+/* Define Fortran mangle from Trilinos macro definition */
+#ifndef F77_BLAS_MANGLE                                                                                       
+# define F77_BLAS_MANGLE@F77_BLAS_MANGLE@                                                                     
+#endif 
+
+/* Define if fortran blas 1 function can return complex type */
+#cmakedefine KOKKOSKERNELS_TPL_BLAS_RETURN_COMPLEX
```
Comment (ndellingwood) [inline cmake/KokkosKernels_config.h.in]: @jjwilke adding you here to keep your work on #120 in sync with a couple updates that were made for CMake builds, this is the matching cmake variable to the commented code above.

**Pull request merged**